metadata-operation-context/src/main/java/io/datahubproject/metadata/services/SecretService.java

```diff
@@ -11,54 +12,140 @@
+
+      // Combine IV and ciphertext
+      ByteBuffer buffer = ByteBuffer.allocate(iv.length + ciphertext.length);
+      buffer.put(iv);
+      buffer.put(ciphertext);
+
+      // Return versioned and encoded result
+      return VERSION_PREFIX + _encoder.encodeToString(buffer.array());
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to encrypt value using AES-GCM!", e);
+    }
+  }
+
+  /** Decrypts a value, supporting both new AES-GCM and legacy AES encryption. */
+  public String decrypt(String encryptedValue) {
+    if (encryptedValue == null) {
+      throw new IllegalArgumentException("Encrypted value cannot be null");
+    }
+
+    // Check if this is new format
+    if (!isLegacyFormat(encryptedValue)) {
+      return decryptV2(encryptedValue.substring(VERSION_PREFIX.length()));
+    } else {
+      // Legacy format - use legacy decryption
+      return decryptLegacy(encryptedValue);
+    }
+  }
+
+  /** Decrypts using the new AES-GCM method. */
+  private String decryptV2(String encryptedValue) {
+    try {
+      byte[] combined;
       try {
-        key = _secret.getBytes(StandardCharsets.UTF_8);
-        sha = MessageDigest.getInstance("SHA-1");
-        key = sha.digest(key);
-        key = Arrays.copyOf(key, 16);
-        secretKey = new SecretKeySpec(key, "AES");
-      } catch (NoSuchAlgorithmException e) {
-        e.printStackTrace();
+        combined = _decoder.decode(encryptedValue);
+      } catch (IllegalArgumentException e) {
+        throw new RuntimeException("Invalid Base64 encoding in encrypted value", e);
       }
-      Cipher cipher = Cipher.getInstance("AES");
-      cipher.init(Cipher.ENCRYPT_MODE, secretKey);
-      return _encoder.encodeToString(cipher.doFinal(value.getBytes(StandardCharsets.UTF_8)));
+
+      // Validate minimum length
+      if (combined.length < GCM_IV_LENGTH + 16) { // IV + min ciphertext + tag
+        throw new RuntimeException("Encrypted value too short for AES-GCM format");
+      }
+
+      // Extract IV and ciphertext
+      ByteBuffer buffer = ByteBuffer.wrap(combined);
+
+      byte[] iv = new byte[GCM_IV_LENGTH];
+      buffer.get(iv);
+
+      byte[] ciphertext = new byte[buffer.remaining()];
+      buffer.get(ciphertext);
+
+      // Initialize cipher
+      Cipher cipher = Cipher.getInstance(AES_GCM_ALGORITHM);
+      GCMParameterSpec gcmSpec = new GCMParameterSpec(GCM_TAG_LENGTH, iv);
+      cipher.init(Cipher.DECRYPT_MODE, _derivedKey, gcmSpec);
+
+      // Decrypt
+      byte[] plaintext = cipher.doFinal(ciphertext);
+      return new String(plaintext, StandardCharsets.UTF_8);
     } catch (Exception e) {
-      throw new RuntimeException("Failed to encrypt value using provided secret!", e);
+      throw new RuntimeException("Failed to decrypt value using AES-GCM!", e);
     }
   }
 
-  public String decrypt(String encryptedValue) {
+  /** Legacy decryption method for backward compatibility. */
+  private String decryptLegacy(String encryptedValue) {
```

Comment:
Can we make the use of decryptLegacy() only available if explicitly enabled? e.g. if you have a default installation of DataHub with v2 keys you shouldn't be able to use legacy mode. Only legacy mode should be available for existing installations and probably have a step in the upgrade or external to migrate the keys to v2.

Thanks!

Reply:
Probably also makes sense to make a docs update encouraging/demanding that people upgrade their secrets so that we can eventually move to disabling the legacy path by default.